File: javascript/core/src/engine/verbs/recode.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */
import { escape, op } from 'arquero'

import type { RecodeArgs } from '../../types.js'
import { makeStepFunction, makeStepNode, wrapColumnStep } from '../factories.js'

/**
 * Executes an arquero derive to map a list of values to new values.
 * Commonly used for recategorization.
 */
const doRecode = wrapColumnStep<RecodeArgs>((input, { column, to, map }) =>
	input.derive({
		[to]: escape((d: any) => op.recode(d[column], map)),
	}),
)

export const recode = makeStepFunction(doRecode)
export const recodeNode = makeStepNode(doRecode)

File: python/datashaper/datashaper/engine/verbs/recode.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#
"""Recode verb implementation."""

from typing import Any

from datashaper.engine.verbs.verb_input import VerbInput
from datashaper.engine.verbs.verbs_mapping import verb
from datashaper.table_store.types import VerbResult, create_verb_result
from datashaper.verbs import recode


@verb(name="recode")
def recode_verb(
    input: VerbInput,
    **kwargs: Any,
) -> VerbResult:
    """Recode verb implementation."""
    result = recode(input.get_input(), **kwargs)
    return create_verb_result(result)
